src/index.ts

```diff
@@ -0,0 +1,326 @@
+    const recencyMap = new Map<string, number>();
+
+    this.recentlyUsedBlocks.forEach((key, index) => {
+      if (!recencyMap.has(key)) {
+        recencyMap.set(key, index + 1);
+      }
+    });
+
+    uniqueRecentBlocks.sort((a, b) => recencyMap.get(a) - recencyMap.get(b));
+
+    return this.generateBlockData(uniqueRecentBlocks);
+  };
+
+  /**
+   * Converts a list of block types to a full-fledge list of block data.
+   *
+   * @param blockTypeList the list of block types
+   * @returns the block data list
+   */
+  generateBlockData = (
+    blockTypeList: string[],
+  ):
+    | Blockly.utils.toolbox.BlockInfo[]
+    | Array<{kind: string; text: string}> => {
+    if (blockTypeList.length === 0) {
+      return [
+        {
+          kind: 'LABEL',
+          text: 'No blocks have been used yet!',
+        },
+      ];
+    }
+
+    return blockTypeList.slice(0, this.numBlocksPerCategory).map((key) => {
+      const json = this.defaultJsonForBlockLookup[key] || {};
+      return {
+        ...json,
+        kind: 'BLOCK',
+        type: key,
+        x: undefined,
+        y: undefined,
+      };
+    });
+  };
+
+  /**
+   * Event listener for workspace events.
+   *
+   * @param e - The event object for workspace events
+   */
+  eventListener(e: Blockly.Events.Abstract): void {
+    if (e.type === Blockly.Events.FINISHED_LOADING) {
+      this.workspaceHasFinishedLoading = true;
+      return;
+    }
+
+    if (
+      e.type === Blockly.Events.BLOCK_CREATE &&
+      this.workspaceHasFinishedLoading &&
+      (e as Blockly.Events.BlockCreate).json?.type
+    ) {
+      const newBlockType = (e as Blockly.Events.BlockCreate).json.type;
+
+      // If this is the first time creating this block, store its default
+      // configuration so we know how exactly to render it in the toolbox.
+      if (!this.defaultJsonForBlockLookup[newBlockType]) {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        this.defaultJsonForBlockLookup[newBlockType] = (e as any).json;
+      }
+      this.recentlyUsedBlocks.unshift(newBlockType);
+    }
+  }
+  /**
+   * Saves the state of this object to a serialized JSON.
+   *
+   * @returns a serialized data object including this object's state
+   */
+  saveToSerializedData(): object {
+    return {
+      recentlyUsedBlocks: this.recentlyUsedBlocks,
+      defaultJsonForBlockLookup: this.defaultJsonForBlockLookup,
+    };
+  }
+
+  /**
+   * Loads the state of this object from a serialized JSON.
+   *
+   * @param state the serialized data payload to load from
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  loadFromSerializedData(state: any): void {
+    if (state.recentlyUsedBlocks) {
+      this.recentlyUsedBlocks = state.recentlyUsedBlocks;
+    }
+    if (state.defaultJsonForBlockLookup) {
+      this.defaultJsonForBlockLookup = state.defaultJsonForBlockLookup;
+    }
+  }
```

Comment:
Remove `any`. Instead declare the parameter type to be an object with properties for `recentlyUsedBlocks` and `defaultJsonForBlockLookup`. Then the `if` statements become unnecessary as you're telling the compiler that the input passed to this function will definitively have these properties. Again, rearrange these properties as was what was present in the original index.js file. Also, make `saveToSerializeData` appear after `loadFromSerializedData`. Again, we want to stay as close to the original `index.js` as possible. This will make it clearer for anyone doing a comparison of the two plugins and make it less likely to introduce new bugs.